openhands/cli/tui.py

```diff
@@ -183,26 +187,45 @@ def display_initial_user_prompt(prompt: str) -> None:
 
 
 # Prompt output display functions
+def display_thought_if_new(thought: str) -> None:
+    """Display a thought only if it hasn't been displayed recently."""
+    global recent_thoughts
+    if thought and thought.strip():
+        # Check if this thought was recently displayed
+        if thought not in recent_thoughts:
+            display_message(thought)
+            recent_thoughts.append(thought)
+            # Keep only the most recent thoughts
+            if len(recent_thoughts) > MAX_RECENT_THOUGHTS:
+                recent_thoughts.pop(0)
+
+
 def display_event(event: Event, config: OpenHandsConfig) -> None:
     global streaming_output_text_area
     with print_lock:
-        if isinstance(event, Action):
-            if hasattr(event, 'thought'):
+        if isinstance(event, CmdRunAction):
+            # For CmdRunAction, display thought first, then command
+            if hasattr(event, 'thought') and event.thought:
                 display_message(event.thought)
-            if hasattr(event, 'final_thought'):
-                display_message(event.final_thought)
-        if isinstance(event, MessageAction):
-            if event.source == EventSource.AGENT:
-                display_message(event.content)
 
-        if isinstance(event, CmdRunAction):
             # Only display the command if it's not already confirmed
             # Commands are always shown when AWAITING_CONFIRMATION, so we don't need to show them again when CONFIRMED
             if event.confirmation_state != ActionConfirmationStatus.CONFIRMED:
                 display_command(event)
 
             if event.confirmation_state == ActionConfirmationStatus.CONFIRMED:
                 initialize_streaming_output()
+        elif isinstance(event, Action):
+            # For other actions, display thoughts normally
+            if hasattr(event, 'thought') and event.thought:
+                display_message(event.thought)
+            if hasattr(event, 'final_thought') and event.final_thought:
+                display_message(event.final_thought)
+
+        if isinstance(event, MessageAction):
+            if event.source == EventSource.AGENT:
+                # Check if this message content is a duplicate thought
+                display_thought_if_new(event.content)
```

Comment:
It seems as though this will get double processed. `display_message(event.thought)` already ran for Action, a superset of action, and the deduping doesn't apply to it.